apps/rush-lib/src/logic/DependencyAnalyzer.ts

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as semver from 'semver';
+import { CommonVersionsConfiguration } from '../api/CommonVersionsConfiguration';
+import { DependencyType, PackageJsonDependency } from '../api/PackageJsonEditor';
+import { RushConfiguration } from '../api/RushConfiguration';
+import { RushConfigurationProject } from '../api/RushConfigurationProject';
+
+export interface IDependencyAnalysis {
+  /**
+   * The common versions configuration from the repo's rush configuration.
+   */
+  commonVersionsConfiguration: CommonVersionsConfiguration;
+
+  /**
+   * A map of all direct dependencies that only have a single semantic version specifier,
+   * unless the variant has the {@link CommonVersionsConfiguration.implicitlyPreferredVersions} option
+   * set to `false`.
+   */
+  implicitlyPreferredVersionByPackageName: Map<string, string>;
+
+  /**
+   * A map of dependency name to the set of version specifiers used in the repo.
+   */
+  allVersionsByPackageName: Map<string, Set<string>>;
+}
+
+export class DependencyAnalyzer {
+  private static _dependencyAnalyzerByRushConfiguration:
+    | Map<RushConfiguration, DependencyAnalyzer>
+    | undefined;
+
+  private _rushConfiguration: RushConfiguration;
+  private _analysisByVariant: Map<string, IDependencyAnalysis> = new Map();
+
+  private constructor(rushConfiguration: RushConfiguration) {
+    this._rushConfiguration = rushConfiguration;
+  }
+
+  public static forRushConfiguration(rushConfiguration: RushConfiguration): DependencyAnalyzer {
+    if (!DependencyAnalyzer._dependencyAnalyzerByRushConfiguration) {
+      DependencyAnalyzer._dependencyAnalyzerByRushConfiguration = new Map();
```

Comment:
If we don't need to be able to enumerate the collection, may as well use a WeakMap for better cleanup.